effekt/shared/src/main/scala/effekt/Namer.scala

```diff
@@ -877,6 +902,19 @@ object Namer extends Phase[Parsed, NameResolved] {
       case Nil => "{}"
       case many => many.mkString("{", ", ", "}")
     }
+
+  /**
+   * Generate a fresh hole id that is unique in the current module.
+   * Ideally, this id should be somewhat stable wrt. edits to the source code.
+   * For this, we use the name of the current top-level definition, disambiguated by a counter that increments in
+   * depth-first order as Namer traverses the syntax tree.
+   */
+  private def freshHoleId: String = {
```

Comment:
Good enough for a start? Better suggestions?

Reply:
The biggest problem I can see is that this is not stable wrt operations we plan to do: taking a hole/goal and splitting it into many smaller holes/goals.
```scala
def addOne(list) = <> // new ~> 1
val someOtherGoal = <> // 2
// ~~>
def addOne(list) = list match {
  case Nil() => <> // new ~> 1
  case Cons(_,_) => <> // new ~> 2
}
val someOtherGoal = <> // was 2 ~> 3
```
So it's impossible to cache some answers/filters/... for `someOtherGoal` as the ID of its hole went from 2 to 3.

Reply:
Good point!

So there are two things that could happen to the document that invalidate the cache

1. user edits the source file (maybe adds new holes, maybe removes some, maybe moves code)
2. we perform some hole filling (removing one hole but replacing it by potentially multiple ones)

I still would go with this and then refine again; let's have it merged ASAP so that @timsueberkrueb can continue.

Reply:
This is probably a tradeoff as to how complex the computation of numbers is and under which operations we want to be stable. 🤔 

We could of course track all definitions (not just top-level ones) and have it be stable here, but this would break once it's not a local def.

We could in principle use the "path" in the tree after that (so body of the first val in ...), which would be stable under "replacing a hole with stuff", but also more effort to implement (and would probably need some "compression" to not have very long names).

Reply:
For the record, I didn't write this as a gotcha, but I just wanted to document that there was a trade-off so that there is _some_ actual record of the potential problems this might introduce in the future.
I strongly believe we forget such things all of the time, leading to problems in the future, because nobody bothers to write them down, hence the comment... :)